### ISSUE TITLE
Firebase Messaging plugin: onLaunch is called multiple times

### ISSUE THREAD

**Issue by ehhc** [flutter/flutter#32698]
Hi,
My app uses firebase cloud messaging for push notifications. Unfortunately, onLaunch is called too often. 
You’ll find an example app to reproduce the issue below (unfortunately, you have to follow all steps to configure the messaging plugin like state here (https://pub.dev/packages/firebase_messaging#-readme-tab-) before you can reproduce the issue).
Following scenario:
The app has two pages. If a push notification is received the app should open page 2. On Page two is a button to go back to page 1. After the user clicks on this button, the app is expected to be on page 1. This works fine, if the app wasn’t terminated when the notification was received. Alas, if it was terminated, the app opens page 1 just for a few minutes and jumps forward to page 2 again.

Steps to reproduce:
1)	Start the app and note the <FCM TOKEN>
2)	Terminate the app
3)	Send a cloud push message to the noted token
4)	--> A notification should appear
5)	Click on the notification
6)	--> Apps opens page 2
7)	Click Button “Back to page 1”
8)	--> <<Expected:>> App is on page 1
9)	--> **<<Actual:>> App is on page 1 for a few milliseconds and opens page 2 again**

## Logs
Unfortunately, logs aren’t possible because is only happens, if the app was terminated previously…
```
C:\Users\x\Projekte\flutter_test\notification\flutter_app>flutter doctor -v
[√] Flutter (Channel beta, v1.5.4-hotfix.2, on Microsoft Windows [Version 10.0.17763.437], locale de-DE)
    • Flutter version 1.5.4-hotfix.2 at C:\Dev\flutter
    • Framework revision 7a4c33425d (2 weeks ago), 2019-04-29 11:05:24 -0700
    • Engine revision 52c7a1e849
    • Dart version 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)


[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\x\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[√] Android Studio (version 3.4)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 35.3.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[√] Connected device (1 available)
    • Pixel 2 • xxxxx • android-arm64 • Android 9 (API 28)

```

## Example App:
```main.dart

import 'package:flutter/material.dart';
import 'package:firebase_messaging/firebase_messaging.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final FirebaseMessaging _firebaseMessaging = FirebaseMessaging();

  @override
  void initState() {
    _firebaseMessaging.getToken().then((token) {
      print(token);
    });
    _firebaseMessaging.configure(
      onMessage: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
      onResume: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
      onLaunch: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
    );
    super.initState();
  }

  _redirectToPageTwo(){
    Navigator.pushReplacement(context, MaterialPageRoute(builder: (context) => PageTwo()));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(child: Text('Page1'),),
    );
  }
}

class PageTwo extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: <Widget>[
          Center(child: Text('Page2'),),
          RaisedButton(child: Text('Go back to page 1'), onPressed: () => Navigator.pushReplacement(context, MaterialPageRoute(builder: (context) => MyHomePage())),)
        ],
      ),
    );
  }
  
}
```

```pubspec.yaml

name: flutter_app
description: A new Flutter application.

version: 1.0.0+1

environment:
  sdk: ">=2.1.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter


dev_dependencies:
  flutter_test:
    sdk: flutter
  firebase_core: ^0.4.0
  firebase_messaging: 5.0.0


flutter:

  uses-material-design: true

```

**Comment by ehhc** [comment]
my issue might be a duplicate of https://github.com/flutter/flutter/issues/28581 . Unfortunately, there is no solution either.. :(

**Comment by kunit1** [comment]
If I remember correctly, this happens because when the Widget that configures the callbacks is popped and then inserted again it re-configures the callbacks and it fires again. As a workaround, you configure the callbacks on the root widget and make sure to not pop or replace that widget. Hence, the initState won't fire and you will keep one callback figuration alive.

**Comment by ehhc** [comment]
@kunit1 thanks for your answer.
Unfortunately, if i try to configure it in the main app (see below), the redirect to the second page isn't working any more (no idea why... the method gets executed..)


```
class _MyAppState extends State<MyApp> {
  final FirebaseMessaging _firebaseMessaging = FirebaseMessaging();

  _redirectToPageTwo(){
    print('message received');
    Navigator.pushReplacement(context, MaterialPageRoute(builder: (context) => PageTwo()));
  }


  @override
  Widget build(BuildContext context) {
    _firebaseMessaging.getToken().then((token) {
      print(token);
    });
    _firebaseMessaging.configure(
      onMessage: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
      onResume: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
      onLaunch: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
    );

    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}
```

**Comment by ehhc** [comment]
Furtheremore, why does a reconfigure also triggeres the method again?

**Comment by ehhc** [comment]
maybe one of the firebase messaging developers can help?
@collinjackson @goderbauer ?

**Comment by ehhc** [comment]
Hi guys, maybe somebody else comes up to this and want to know whether i found a solution.. I worked around the problem with the following:
Instead of `Navigator.pushReplacement([...])` i use a `popUntil`-mechanism to navigate between the pages. That way, the page which initializes the push notification handler won't be re-created and therefore dosen't triggers the notification handling again.

You can find my code below (and can compare it to the initial code at the top of this issue):

```
import 'package:flutter/material.dart';
import 'package:firebase_messaging/firebase_messaging.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      initialRoute: "/",
      routes: {
        '/': (context) => MyHomePage(),
        '/home': (context) => MyHomePage(),
        '/pTwo': (context) => PageTwo(),
      },
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final FirebaseMessaging _firebaseMessaging = FirebaseMessaging();

  @override
  void initState() {
    _firebaseMessaging.getToken().then((token) {
      print(token);
    });
    _firebaseMessaging.configure(
      onMessage: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
      onResume: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
      onLaunch: (Map<dynamic, dynamic> message) => _redirectToPageTwo(),
    );
    super.initState();
  }

  _redirectToPageTwo(){
    Navigator.pushNamed(context, "/pTwo");
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(child: Text('Page1'),),
    );
  }
}

class PageTwo extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: <Widget>[
          Center(child: Text('Page2'),),
          RaisedButton(child: Text('Go back to page 1'), onPressed: () async {
            Navigator.popUntil(context, ModalRoute.withName('/'));
          },)
        ],
      ),
    );
  }
}
```


As i said, the important part is the usage of `Navigator.popUntil(context, ModalRoute.withName('/'));` instead of `Navigator.pushReplacement(context, MaterialPageRoute(builder: (context) => MyHomePage()))` 

Please let me know if you find a better solution :)


Btw: maybe the documentation should be updated to point to the problem of this issue?

**Comment by Vikas-Jilla** [comment]
instead of calling `Navigator.pushReplacement(..)` you should just call `Navigator.of(context).push(router)` and `Navigator.of(context).pop(result)` -> when back is clicked

**Comment by BondarenkoStas** [comment]
Thank you @Vikas-Jilla @dkobia ! @ehhc did you manage to solve it with comment above?

**Comment by mugikhan** [comment]
I'm having the same issue. The onLaunch runs everythime I load the initial route where the configure code is. I don't know how to stop this behaviour.

**Comment by Minituff** [comment]
Same issues as well. The only way I was able to (kind of) work around it was to run fcm.configure() during initState in MyApp. As far as I know MyApp should only be initialized once and then onLaunch shoudnt run again after that. But during development if I ever perform a hot-restart then initState runs again and onLaunch runs again. Is there an fcm.clear() or anything like that? I feel like there should be a way to let fcm know that I've already handled that message.

**Comment by akshatjagga** [comment]
This has become an issue because I load my data in FutureBuilder so configuring fcm in init doesn't work because I need that data to navigate to a specified page. No fix till now :(

**Comment by kroikie** [comment]
@ehhc

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/1060. Any further collaboration will be done there.

**Comment by ehhc** [comment]
> 
> 
> Thank you @Vikas-Jilla @dkobia ! @ehhc did you manage to solve it with comment above?

No, unfortunately this does not solve my problem. there are reasons why i need to call `pushReplacement` so avaoiding it is not a working solution for me.



> 
> 
> @ehhc
> 
> This issue has been moved to [FirebaseExtended/flutterfire#1060](https://github.com/FirebaseExtended/flutterfire/issues/1060). Any further collaboration will be done there.

Okay. Thanks for pinging me. I'll have a look on whether it will be solved in the future. Hope so :)

**Comment by nemesioo** [comment]
Enclose it with a singleton
class MySingleton {
  static final MySingleton _singleton = MySingleton._internal();
  factory MySingleton() => _singleton;
  
  MySingleton._internal() {
    // fcm.configure
  }
  
  // Methods, variables ...
}

**Comment by LOG-TAG** [comment]
@nemesioo can you share complete code with singleton?

**Comment by Pratikdomadiya** [comment]
hello guys!!!
i have a same issue. when my application was killed , i got a notification from firebase , but it navigate to homeScreen of my app but not on a page which i want. please suggest solution for it.

here is full code...
[notification_code.txt](https://github.com/flutter/flutter/files/4823235/notification_code.txt)

**Comment by Pratikdomadiya** [comment]
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:wholetex_seller/providers/auth.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:wholetex_seller/screens/order_screen.dart';
import 'package:wholetex_seller/widgets/mynavigationbar.dart';
import 'package:wholetex_seller/screens/order_screen.dart';
import 'package:wholetex_seller/widgets/home_screen.dart';
import 'package:wholetex_seller/widgets/mynavigationbar.dart';

class NotificationHandling {
  FlutterLocalNotificationsPlugin flutterLocalNotificationsPlugin;

  void initialize() {
    flutterLocalNotificationsPlugin = new FlutterLocalNotificationsPlugin();
    var android = new AndroidInitializationSettings('@mipmap/ic_launcher');
    var ios = new IOSInitializationSettings();
    var initSettings = new InitializationSettings(android, ios);
    flutterLocalNotificationsPlugin.initialize(initSettings,
        onSelectNotification: onSelectNotification);
  }

  Future onSelectNotification(String payload) async {
    debugPrint("payload : $payload");
    // showDialog(
    //   context: context,
    //   builder: (_) => new AlertDialog(
    //     title: new Text('Notification'),
    //     content: new Text('$payload'),
    //   ),
    // );
  }

  showNotification(Map<String, dynamic> msg) async {
    var android = new AndroidNotificationDetails(
      'sdffds dsffds',
      "CHANNLE NAME",
      "channelDescription",
    );
    var iOS = new IOSNotificationDetails();
    var platform = new NotificationDetails(android, iOS);
    await flutterLocalNotificationsPlugin.show(
        0, msg['notification']['title'], msg['notification']['body'], platform);
  }
  


  void registerFirebaseNotification(BuildContext context) {
    FirebaseMessaging fbm = new FirebaseMessaging();
    initialize();
    fbm.configure(
      onLaunch: (Map<String, dynamic> msg) async {
        print(" onLaunch called ${(msg)}");
        if (msg['data ']['screen'] == 'order_page') {
          print(msg['data']['screen']);
          // await navigatorKey.currentState.push(
          //     MaterialPageRoute(builder: (context) => MyNavigationBar(2)));
          await Navigator.pushReplacement(
            context,
            MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          );
          await Navigator.pushReplacement(
            context,
            MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          );
        }
      },
      onResume: (Map<String, dynamic> msg) async {
        print(" onResume called ${(msg)}");
        print(msg['data']['screen']);
        if (msg['data']['screen'] == 'order_page') {
          print(msg['data']['screen']);
          // await navigatorKey.currentState.pushReplacement(
          //     MaterialPageRoute(builder: (context) => MyNavigationBar(2)));
          await Navigator.push(
            context,
            MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          );
          print(msg['data']['screen']);
        }
      },
      onMessage: (Map<String, dynamic> msg) async {
        //showNotification(msg);
        print(" onMessage called ${(msg)}");
        if (msg['data']['screen'] == 'order_page') {
          print(msg['data']['screen']);
          showNotification(msg);
          //myBackgroundMessageHandler(msg, context);
          // navigatorKey.currentState.pushReplacement(
          //     MaterialPageRoute(builder: (context) => MyNavigationBar(2)));
          // await Navigator.push(
          //   context,
          //   MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          // );
        }
      },
    );
    fbm.requestNotificationPermissions(
        const IosNotificationSettings(sound: true, alert: true, badge: true));
    fbm.onIosSettingsRegistered.listen((IosNotificationSettings setting) {
      print('IOS Setting Registed');
    });
    fbm.getToken().then((token) {
      update(token);
    });
  }

  Future<dynamic> myBackgroundMessageHandler(
      Map<String, dynamic> message, BuildContext context) async {
    print("onLaunch: $message");
    return showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('sfsdfdsfdf'),
        content: Text('contect'),
      ),
    );
    //_showBigPictureNotification(message);
    // Or do other work.
  }

  update(String token) async {
    SharedPreferences localStorage = await SharedPreferences.getInstance();
    String userToken = localStorage.getString('token');
    // userToken = userToken.replaceAll('"', '');

    print(userToken);

    if (userToken != token) {
      Auth().saveSellerToken(token);
    }
    print('token:' + token);

    //DatabaseReference databaseReference = new FirebaseDatabase().reference();
    //databaseReference.child('fcm-token/${token}').set({"token": token});
    //textValue = token;
    //setState(() {});
  }
}

**Comment by Pratikdomadiya** [comment]
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:wholetex_seller/providers/auth.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:wholetex_seller/screens/order_screen.dart';
import 'package:wholetex_seller/widgets/mynavigationbar.dart';
import 'package:wholetex_seller/screens/order_screen.dart';
import 'package:wholetex_seller/widgets/home_screen.dart';
import 'package:wholetex_seller/widgets/mynavigationbar.dart';

class NotificationHandling {
  FlutterLocalNotificationsPlugin flutterLocalNotificationsPlugin;

  void initialize() {
    flutterLocalNotificationsPlugin = new FlutterLocalNotificationsPlugin();
    var android = new AndroidInitializationSettings('@mipmap/ic_launcher');
    var ios = new IOSInitializationSettings();
    var initSettings = new InitializationSettings(android, ios);
    flutterLocalNotificationsPlugin.initialize(initSettings,
        onSelectNotification: onSelectNotification);
  }

  Future onSelectNotification(String payload) async {
    debugPrint("payload : $payload");
    // showDialog(
    //   context: context,
    //   builder: (_) => new AlertDialog(
    //     title: new Text('Notification'),
    //     content: new Text('$payload'),
    //   ),
    // );
  }

  showNotification(Map<String, dynamic> msg) async {
    var android = new AndroidNotificationDetails(
      'sdffds dsffds',
      "CHANNLE NAME",
      "channelDescription",
    );
    var iOS = new IOSNotificationDetails();
    var platform = new NotificationDetails(android, iOS);
    await flutterLocalNotificationsPlugin.show(
        0, msg['notification']['title'], msg['notification']['body'], platform);
  }
  


  void registerFirebaseNotification(BuildContext context) {
    FirebaseMessaging fbm = new FirebaseMessaging();
    initialize();
    fbm.configure(
      onLaunch: (Map<String, dynamic> msg) async {
        print(" onLaunch called ${(msg)}");
        if (msg['data ']['screen'] == 'order_page') {
          print(msg['data']['screen']);
          // await navigatorKey.currentState.push(
          //     MaterialPageRoute(builder: (context) => MyNavigationBar(2)));
          await Navigator.pushReplacement(
            context,
            MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          );
          await Navigator.pushReplacement(
            context,
            MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          );
        }
      },
      onResume: (Map<String, dynamic> msg) async {
        print(" onResume called ${(msg)}");
        print(msg['data']['screen']);
        if (msg['data']['screen'] == 'order_page') {
          print(msg['data']['screen']);
          // await navigatorKey.currentState.pushReplacement(
          //     MaterialPageRoute(builder: (context) => MyNavigationBar(2)));
          await Navigator.push(
            context,
            MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          );
          print(msg['data']['screen']);
        }
      },
      onMessage: (Map<String, dynamic> msg) async {
        //showNotification(msg);
        print(" onMessage called ${(msg)}");
        if (msg['data']['screen'] == 'order_page') {
          print(msg['data']['screen']);
          showNotification(msg);
          //myBackgroundMessageHandler(msg, context);
          // navigatorKey.currentState.pushReplacement(
          //     MaterialPageRoute(builder: (context) => MyNavigationBar(2)));
          // await Navigator.push(
          //   context,
          //   MaterialPageRoute(builder: (context) => MyNavigationBar(2)),
          // );
        }
      },
    );
    fbm.requestNotificationPermissions(
        const IosNotificationSettings(sound: true, alert: true, badge: true));
    fbm.onIosSettingsRegistered.listen((IosNotificationSettings setting) {
      print('IOS Setting Registed');
    });
    fbm.getToken().then((token) {
      update(token);
    });
  }

  Future<dynamic> myBackgroundMessageHandler(
      Map<String, dynamic> message, BuildContext context) async {
    print("onLaunch: $message");
    return showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('sfsdfdsfdf'),
        content: Text('contect'),
      ),
    );
    //_showBigPictureNotification(message);
    // Or do other work.
  }

  update(String token) async {
    SharedPreferences localStorage = await SharedPreferences.getInstance();
    String userToken = localStorage.getString('token');
    // userToken = userToken.replaceAll('"', '');

    print(userToken);

    if (userToken != token) {
      Auth().saveSellerToken(token);
    }
    print('token:' + token);

    //DatabaseReference databaseReference = new FirebaseDatabase().reference();
    //databaseReference.child('fcm-token/${token}').set({"token": token});
    //textValue = token;
    //setState(() {});
  }
}

**Comment by geecat** [comment]
> hello guys!!!
> i have a same issue. when my application was killed , i got a notification from firebase , but it navigate to homeScreen of my app but not on a page which i want. please suggest solution for it.
> 
> here is full code...
> [notification_code.txt](https://github.com/flutter/flutter/files/4823235/notification_code.txt)

Please post your code here?
Why are you attaching a code in text file? Use pastebin or someother.

**Comment by Strain-solutions** [comment]
Hi all - leaving this here in case it helps debug / work around issues.  I am having the same issue with onLaunch or onResume called repeatedly on android.  In my app clicking the notification takes you to a specific 'notifications screen' where images, urls etc are processed.  

Originally this screen used a floating action button to navigate back to the homepage (using  standard

   `Navigator.push(context,
                              MaterialPageRoute(builder: (context) =>
                                  HomeScreen()));`

This would take the user back to the homepage and re-trigger the messages screen resulting in loop the user can't escape.

However, if I include the standard app bar including the leading navigator arrow (to go back to the previous screen) then the loop is NOT triggered.  The user goes back the the home screen without the messages screen being started.

In fact currently for testing I have both FAB and appBar on the message screen and the appBar doesn't trigger the loop whereas the FAB button does FOR THE SAME MESSAGE

Im pretty new to flutter so kind of at the limits of my knowledge here, but is there something different between the appBar navigator and the MaterialPageRoute navigation that would explain the behaviour?

**Comment by hx269988653** [comment]
When the application terminates, there is no notification explaining how to configure it. The foreground and background can receive onMessage and onResume.

**Comment by Bader-Al** [comment]
> I had the same issue until I realized that this might all depend on how `firebaseMessaging.configure` is done. It seems to me that as @kunit1 mentioned, depending on where you perform this configuration, it is probably getting reconfigured again. Either put it in the root or potentially put it in a singleton as I did and configure it only once re:
> 
> This worked for me.
> 
> ```
> bool _isConfigured = false;
> 
> ...
> 
> if (!_isConfigured) {
>   _firebaseMessaging.configure(
>     onMessage: (Map<String, dynamic> message) {
>       print('onMessage: '+message.toString());
>       return;
>     },
>     onResume: (Map<String, dynamic> message) {
>       print('onResume: '+message.toString());
>       return;
>     },
>     onLaunch: (Map<String, dynamic> message) {
>       print('onLaunch: '+message.toString());
>       return;
>     },
>   );
>   _isConfigured = true;
> }
> ```

This answer worked for me **only when I declared `isConfigured = false` ** as static. 

You may need to do this if you enclosed this in a notificationsService class constructed down the widget tree. 
following that, I set it to true as such
`PushNotificationsService.isConfigured = true;` . This way it stays true even after rebuild since it's a static variable (It's independent of object instantiation)
